src/main/scala/is/hail/methods/SampleQC.scala

```diff
@@ -48,76 +48,76 @@ object SampleQCCombiner {
     "rTiTv" -> TFloat64,
     "rHetHomVar" -> TFloat64,
     "rInsertionDeletion" -> TFloat64)
+
+  val ti = 0
+  val tv = 1
+  val ins = 2
+  val del = 3
+  val star = 4
+
+  def alleleIndices(v: Variant): Array[Int] = {
+    val alleleTypes = v.altAlleles.map { a =>
+      a.altAlleleType match {
+        case AltAlleleType.SNP => if (a.isTransition) ti else tv
+        case AltAlleleType.Insertion => ins
+        case AltAlleleType.Deletion => del
+        case _ => -1
+      }
+    }.toArray
+
+    alleleTypes
+  }
 }
 
-class SampleQCCombiner(val keepStar: Boolean) extends Serializable {
+class SampleQCCombiner extends Serializable {
   var nNotCalled: Int = 0
   var nHomRef: Int = 0
   var nHet: Int = 0
   var nHomVar: Int = 0
 
-  var nSNP: Int = 0
-  var nIns: Int = 0
-  var nDel: Int = 0
+  val aCounts: Array[Int] = Array.fill[Int](5)(0)
   var nSingleton: Int = 0
-  var nTi: Int = 0
-  var nTv: Int = 0
 
   val dpSC: StatCounter = new StatCounter()
 
   val gqSC: StatCounter = new StatCounter()
 
-  // FIXME per-genotype
+  def merge(aTypes: Array[Int], acs: Array[Int], view: HTSGenotypeView): SampleQCCombiner = {
 
-  def merge(v: Variant, acs: Array[Int], g: Genotype): SampleQCCombiner = {
-
-    val gt = Genotype.unboxedGT(g)
-    if (gt < 0)
-      nNotCalled += 1
-    else {
-      if (gt == 0) {
+    if (view.hasGT) {
+      val gt = view.getGT
+      if (gt == 0)
         nHomRef += 1
-      } else {
+      else {
         val gtPair = Genotype.gtPair(gt)
-
-        def mergeAllele(ai: Int) {
-          if (ai > 0 && (keepStar || !v.altAlleles(ai - 1).isStar)) {
-
-            val altAllele = v.altAlleles(ai - 1)
-            altAllele.altAlleleType match {
-              case AltAlleleType.SNP =>
-                nSNP += 1
-                if (altAllele.isTransition)
-                  nTi += 1
-                else
-                  nTv += 1
-              case AltAlleleType.Insertion =>
-                nIns += 1
-              case AltAlleleType.Deletion =>
-                nDel += 1
-              case _ =>
-            }
-
-            if (acs(ai) == 1)
+        val j = gtPair.j
+        val k = gtPair.k
+
+        def mergeAllele(idx: Int) {
+          if (idx > 0) {
+            val aType = aTypes(idx - 1)
+            aCounts(aType) += 1
```

Comment:
I believe this will fail if aType is -1 as is the case for Complex, MNP, or Star.